src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -302,9 +302,10 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
                     continue;
                 }
 
-                (var sources, var diagnostics) = context.ToImmutableAndFree();
-                stateBuilder[i] = new GeneratorState(generatorState.Info, generatorState.PostInitTrees, ParseAdditionalSources(generator, sources, cancellationToken), diagnostics);
-                diagnosticsBag?.AddRange(diagnostics);
+                (var sources, var generatorDiagnostics) = context.ToImmutableAndFree();
```

Comment:
it would be nice if we could avoid the allocation (just allocate the final array of filtered diagnostics, not the intermediate one), but I am not extremely concerned about it.

Reply:
@rikkigibson Yeah I actually thought about this, but with the upcoming work we'll be collecting diagnostics from multiple places. I figured it made sense to centralize the filtering upfront now, rather than having it spread across multiple places down the line.